Fixes Issue #204

haidangwa opened the pull request:
# Description

Fixes the splunkd command line arguments used in `#splunk_login_successful?` helper method

## Issues Resolved

#204 

## Check List

- [ ] All tests pass. See TESTING.md for details.
- [ ] New functionality includes testing.
- [ ] New functionality has been documented in the README if applicable.

ramereth commented:
@haidangwa can you please rebase instead of doing a merge commit update to keep the history linear? Also please address the chefspec issues.

haidangwa commented:
Working on it now

On Wed, May 12, 2021, 9:41 AM Lance Albertson ***@***.***>
wrote:

> @haidangwa <https://github.com/haidangwa> can you please rebase instead
> of doing a merge commit update to keep the history linear? Also please
> address the chefspec issues.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sous-chefs/chef-splunk/pull/205#issuecomment-839930170>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOYB3JNKABYIRKGWJVB5CLTNKVULANCNFSM4ZHLQ5PA>
> .
>

ramereth commented:
@haidangwa looks like we need to deal with some Chef 17 issues before we can get this merged unfortunately.

haidangwa commented:
@ramereth I'm running the latest chef-workstation on my laptop and the test for uninstall-forwarder passes in Test Kitchen with Chef 17. I'm not sure why this is failing in Github Actions.

ramereth commented:
@haidangwa I can replicate the problem on my end with 17.1.35. I suspect it has something to do with the `systemd?` helper included the cookbook conflicting with something else. That really should use `chef-utils` helpers IMO, but that would increase the Chef Client requirement (which we're going to need to do anyway).

haidangwa commented:
@ramereth my latest change was to remove that dependency on the `systemd?` helper entirely. I replaced it with a `uses_systemd?` helper that is in the resource class. The latter doesn't use the `node` object at all; in fact, it shells out and runs a command to find out if the system is using systemd.

haidangwa commented:
check out https://github.com/sous-chefs/chef-splunk/pull/205/commits/85fa20354fcc5c9a80e95b8773cb1078f684d712

haidangwa commented:
This feels like a bug in Chef 17.1 -- all of the tests pass with Chef Infra Client 17.0.x and the uninstall-forwarder tests fail with Chef Infra Client 17.1.39.

haidangwa commented:
@ramereth I've narrowed this down even further. Using kitchen-dokken, the chef infra clients pass all the way until 17.0.199. It breaks between 17.0.199 and 17.0.200. That seems definitive that it is a Chef Infra Client bug.

haidangwa commented:
looks like ruby 3.0.1 was added to the chef omnibus_overrides.rb between v17.0.197 and v17.0.198

```
diff --git a/omnibus_overrides.rb b/omnibus_overrides.rb
index d20d702d73..32c2cc6e69 100644
--- a/omnibus_overrides.rb
+++ b/omnibus_overrides.rb
@@ -16,7 +16,8 @@ override "ncurses", version: "5.9"
 override "nokogiri", version: "1.11.0"
 override "openssl", version: mac_os_x? ? "1.1.1k" : "1.0.2y"
 override "pkg-config-lite", version: "0.28-1"
-override "ruby", version: "2.7.2"
+override "bundler", version: "2.2.15"
+override "ruby", version: "3.0.1"
 override "ruby-windows-devkit-bash", version: "3.1.23-4-msys-1.0.18"
 override "util-macros", version: "1.19.0"
 override "xproto", version: "7.0.28"
```

haidangwa commented:
I think getting this cookbook to be compatible with Chef Infra Client 17 is beyond the scope of this PR. Fixing the changes to be Chef 17 compatible should be in itself a separate Pull Request.

haidangwa commented:
@ramereth OK. With all that said (above), I am thinking the problem is in the `service` provider/resource; perhaps the recent changes to the service resource in Chef 17. In my custom resource, `splunk_installer`, it calls two helper methods, and both helper methods are referencing the node data.

```
  service 'Splunk' do
    service_name server? ? 'Splunkd' : 'SplunkForwarder'
    action systemd? ? %i(stop disable) : :stop
  end
```
Note: `server?` and `systemd?` are helper methods declared in `libraries/helpers.rb` and both return true/false after making a comparison using a node attribute.

However, the error does not fail for the `server?` and fails for the `systemd?`. It is perhaps a bug in the service provider's `action` argument.

haidangwa commented:
With all the above being said, I have refactored the pull request changes. It is now compatible with Chef 17 (with its alleged bug in the `service` provider) and removed a few unneeded dependencies. I've updated the PR description to match my entries in CHANGELOG.md.

haidangwa commented:
All tests pass under Chef 17. This awaits a review and merge.

ramereth commented on .github/workflows/ci.yml:
```diff
@@ -12,27 +12,27 @@ jobs:
+        uses: actions/checkout@main
       - name: Run Chef Delivery
-        uses: actionshub/chef-delivery@master
+        uses: actionshub/chef-delivery@1.0.0
```
Let's please leave all of these to either `main` (if it exists) or `master`.

haidangwa replied:
IMHO, attaching to master leaves the outcome of the tests in a potentially unpredictable state, as the HEAD of master/main could be potentially unstable. I'd rather leave it at a specific version that has been fully tested and released knowing that the code can be stable.

ramereth replied:
While I agree, we also manage most of these actions and know the state of the main/master branches. We'd like to be able to push any updates that are needed easily to cookbooks instead of having to bump them. Our standard has been to use the branches and not tags on all the other cookbooks.

haidangwa replied:
Then why have releases at all?

ramereth commented on CHANGELOG.md:
```diff
@@ -4,6 +4,15 @@ This file is used to list changes made in each version of the splunk cookbook.
 
 ## Unreleased
 
+- Fixes Issue [#204](https://github.com/sous-chefs/chef-splunk/issues/204)# Please enter the commit message for your changes. Lines starting
```
Looks like you didn't mean to add this.
```suggestion
- Fixes Issue [#204](https://github.com/sous-chefs/chef-splunk/issues/204)#
```

haidangwa replied:
fixed the change log entry with 4aad88c0fefcb2e41065471d6bf3d8394f16ec38

ramereth commented on CHANGELOG.md:
```diff
@@ -4,6 +4,15 @@ This file is used to list changes made in each version of the splunk cookbook.
+- Addresses linting issues
+- Fixes an issue with the `user-seed.conf` file
+- Ensures that splunk is installed prior to anything in the `chef-splunk::service` recipe executes
+- Fixes Chef Delivery job in Github Actions: checks out `main` branch instead of `master` branch
```
This should be using `main` if the upstream branch exists, otherwise they should remain using `master.

haidangwa replied:
As previously commented, IMHO, attaching to master/main leaves the outcome of the tests in a potentially unpredictable state, as the HEAD of master/main could be potentially unstable. I'd rather leave it at a specific version that has been fully tested and released knowing that the code can be stable.

ramereth commented on kitchen.dokken.yml:
```diff
@@ -1,8 +1,9 @@
   name: dokken
   privileged: true  # because Docker and SystemD/Upstart
-  chef_version: <%= ENV['CHEF_VERSION'] || 'current' %>
+  chef_version: <%= ENV['CHEF_VERSION'] || 'latest' %>
```
Please keep this using `current`
```suggestion
  chef_version: <%= ENV['CHEF_VERSION'] || 'current' %>
```

haidangwa replied:
`current` can potentially test against a version of Chef Infra Client that is considered unstable. I'd rather not be testing against unstable/unreleased Chef Infra client versions. `Latest` appears to be a good compromise in terms of getting the newest version possible and also testing against a stable version.

ramereth replied:
This catches any potential upstream issues/fixes before a stable release that happens upstream and this has been a standard we've been using globally on all of our cookbooks. It should be stable enough for testing IMHO.

haidangwa replied:
What you're proposing unnecessarily ties the release of chef-splunk to the releases of other cookbooks. This setting of `current` was the root cause of the initial Chef 17 incompatibility issues. The version of Chef Infra Client that GH Actions was pulling in and testing was different than the version tested when the PR was submitted. The `current` setting allows for "whatever version happens to be on the test machine" at the time of test. In this case, the version was 17.0.150 that was never released -- except only in Docker Hub -- and GH Actions tested with the 17.0.242 version that was an official released version. Setting this as `current` creates an unstable development release cycle.

ramereth commented on kitchen.dokken.yml:
```diff
@@ -1,8 +1,9 @@
-  chef_version: <%= ENV['CHEF_VERSION'] || 'current' %>
+  chef_version: <%= ENV['CHEF_VERSION'] || 'latest' %>
   chef_license: accept-no-persist
+  clean_dokken_sandbox: true
```
Please remove this.

haidangwa replied:
Why? This does not appear to affect the testing environment in Github Actions; however, I've found that having it in place helps with local development, as it keeps the development environment clean while testing/developing as near as possible to the release environment.

ramereth replied:
This hasn't been a standard in our files and I'd rather have it in a state that's similar to other cookbooks. If it's useful locally, then you always use `.kitchen.local.yml` for those types of settings.

haidangwa replied:
KITCHEN_LOCAL_YAML is already set to the `kitchen.dokken.yml` file for this repo -- Test Kitchen will ignore `.kitchen.local.yml` and `kitchen.local.yml`.

ramereth commented on kitchen.dokken.yml:
```diff
@@ -42,6 +43,12 @@ platforms:
+        splunk:
+          server:
+            url: https://download.splunk.com/products/splunk/releases/8.0.9/linux/splunk-8.0.9-153839c8b72f-linux-2.6-x86_64.rpm
+            version: 8.0.9
```
Why is this needed?

haidangwa replied:
This change is noted in the commit message: commit b0b8b5b3b293e5860afea481b870e3a100300a63

```
Author: Dang H. Nguyen <dang.nguyen@disney.com>
Date:   Fri May 14 16:13:02 2021 -0700

    for the shdeployer test, failure was due to CentOS7 RPM corruption from Splunk; therefore, this change will use the 8.0.9 RPM instead
```

Changing this attribute value in the cookbook would be a breaking change, per semver, so leaving the default attribute value the way it is and adding this in the Test Kitchen environment is a decent workaround.

ramereth commented on kitchen.yml:
```diff
@@ -1,6 +1,6 @@
 ---
 provisioner:
-  deprecations_as_errors: true
+  deprecations_as_errors: false
```
We need to keep this on.

haidangwa replied:
This will bring up deprecation messages for upstream dependencies that we have no control over when they will get fixed, if they will get fixed. Leaving this in place will never get a fully passing CI run for all future pull requests.

haidangwa replied:
Below is a screenshot from one of the latest/passing tests in GH Actions. As you can see, the only deprecation messages are for upstream dependencies that we have no control over. This value set to `true` was previously causing the tests to FAIL for no other reason.

haidangwa replied:
![image](https://user-images.githubusercontent.com/1933549/118560886-0b83fa80-b71f-11eb-9881-289a5d033617.png)

ramereth replied:
I know that the line cookbook should have those issues resolved by now. We shouldn't be using the chef-vault or ohai in this cookbook anymore as both of them either are or will be deprecated upstream and will not get fixes for Chef 17.

haidangwa replied:
Although chef-vault is built into Chef Infra Client, it does not have the data bag fallback from the `chef_vault_item` method that this cookbook supports. Until Chef Infra Client supports that in its implementation of chef-vault, we need to have the cookbook as a dependency. There have been some talk about removing the dependency on chef-vault altogether, but until a PR is submitted to do that, this cookbook will need to have the chef-vault cookbook as a dependency.

haidangwa replied:
Unless we remove support for older Chef Infra clients before 17, this cookbook needs some stuff to remain backwards compatible.

haidangwa replied:
Also, v4.0.0 of the line cookbook, which this change also updated to pull in, does not address the deprecated issues. This cookbook should not have to wait for another cookbook to fix their failings in order to be released.

ramereth commented on libraries/helpers.rb:
```diff
@@ -124,7 +124,7 @@ def splunk_runas_user
 
     def splunk_service_provider
-      if node['init_package'] == 'systemd'
+      if systemd?
```
This whole method should go away as we are no longer supporting any distribution that isn't using systemd.

haidangwa replied:
However, even the systems that support systemd can still be running SysV scripts, depending on the targeted deployment envrionment. Splunk installs need to be able to know what environment they're going in. This method needs to stay.

haidangwa replied:
Removing this method in this PR is also way out of scope for this PR. This PR is submitted to address a specific BUG, not to completely rewrite the cookbook -- the latter of which would be a breaking change. Removing the helper method would be a breaking change, not a bugfix.

ramereth replied:
True on being out of scope. The Chef 17 fixes really should happen first in a separate PR and then this PR rebased against that once it's done. That PR could deal with this issue specfically.

haidangwa replied:
Until that happens, this PR is ready to go, IMHO.

ramereth commented on resources/splunk_installer.rb:
```diff
@@ -123,24 +127,28 @@ def download_package
+      action :delete
+      triggers_reload true
+    end
+  end
```
Why are you removing the systemd unit files?

haidangwa replied:
This is in the `remove` action of the resource -- of course, the systemd unit files need to be removed.

ramereth replied:
Doesn't the RPM/DEB package already handle that for us?

haidangwa replied:
It obviously does not, as the sequence in the recipe to remove the systemd unit files is executed AFTER the package is removed. Splunk's installer does not install the systemd unit files, except by a post-install command and it's entirely optional by the deployment environment. Been this way for years -- nothing new here, except improving the way that the files get removed.

ramereth reviewed (changes requested):
Thanks for the improvements!

damacus commented on .github/workflows/ci.yml:
```diff
@@ -12,27 +12,27 @@ jobs:
+        uses: actions/checkout@main
       - name: Run Chef Delivery
-        uses: actionshub/chef-delivery@master
+        uses: actionshub/chef-delivery@1.0.0
```
```suggestion
        uses: actionshub/chef-delivery@main
```

damacus commented on .github/workflows/ci.yml:
```diff
@@ -12,27 +12,27 @@ jobs:
+        uses: actions/checkout@main
       - name: Run yaml Lint
-        uses: actionshub/yamllint@master
+        uses: actionshub/yamllint@1.0.0
```
```suggestion
        uses: actionshub/yamllint@main
```

damacus commented on .github/workflows/ci.yml:
```diff
@@ -12,27 +12,27 @@ jobs:
+        uses: actions/checkout@main
       - name: Run Markdown Lint
-        uses: actionshub/markdownlint@master
+        uses: actionshub/markdownlint@2.0.2
```
```suggestion
        uses: actionshub/markdownlint@main
```

damacus commented on .github/workflows/ci.yml:
```diff
@@ -66,11 +66,11 @@ jobs:
+        uses: actions/checkout@main
       - name: Install Chef
-        uses: actionshub/chef-install@master
+        uses: actionshub/chef-install@2.0.1
```
```suggestion
        uses: actionshub/chef-install@main
```

damacus commented on .github/workflows/ci.yml:
```diff
@@ -66,11 +66,11 @@ jobs:
+        uses: actionshub/chef-install@2.0.1
       - name: Dokken
-        uses: actionshub/kitchen-dokken@master
+        uses: actionshub/kitchen-dokken@1.0.1
```
```suggestion
        uses: actionshub/kitchen-dokken@main
```

damacus commented on kitchen.dokken.yml:
```diff
@@ -1,7 +1,7 @@
   name: dokken
   privileged: true  # because Docker and SystemD/Upstart
-  chef_version: <%= ENV['CHEF_VERSION'] || 'current' %>
+  chef_version: <%= ENV['CHEF_VERSION'] || 'latest' %>
```
```suggestion
  chef_version: <%= ENV['CHEF_VERSION'] || 'current' %>
```

damacus commented on kitchen.yml:
```diff
@@ -1,6 +1,6 @@
 ---
 provisioner:
-  deprecations_as_errors: true
+  deprecations_as_errors: false
```
```suggestion
  deprecations_as_errors: true
```

damacus commented on libraries/helpers.rb:
```diff
@@ -269,7 +269,8 @@ def upgrade_enabled?
     def systemd?
-      node['init_package'] == 'systemd'
+      ps1 = shell_out('ps --no-headers 1')
+      ps1.stdout.strip.match?(/systemd/)
```
Why are we not relying on Ohai to determine the init package here?

haidangwa replied:
I believe this was a workaround put in place when ubuntu 14.04 was still being tested -- back when [this bug](https://github.com/chef/ohai/issues/917) was affecting tests.

damacus commented on test/fixtures/cookbooks/test/metadata.rb:
```diff
@@ -1,6 +1,6 @@
 depends 'chef-splunk'
 depends 'yum-centos', '>= 3.1.0'
-depends 'line', '~> 2.9.1'
+depends 'line', '~> 4.0.0'
```
```suggestion
depends 'line'
```

damacus commented on test/fixtures/cookbooks/test/metadata.rb:
```diff
@@ -1,6 +1,6 @@
 name 'test'
-version '0.0.2'
+version '0.1.0'
```
```suggestion
version '0.0.2'
```

ramereth commented:
@haidangwa so it looks like some of the fixes you have in here were done in #215 and conflicts with this. Can you please rebase and resolve the conflicts? Thanks!
